contracts/modules/royalty/policies/IpRoyaltyVault.sol

```diff
@@ -0,0 +1,237 @@
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice LAP royalty policy address
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    IRoyaltyPolicyLAP public immutable ROYALTY_POLICY_LAP;
+
+    /// @notice Ip id to whom this royalty vault belongs to
+    address public ipId;
+
+    /// @notice Amount of unclaimed royalty tokens
+    uint32 public unclaimedRoyaltyTokens;
+
+    /// @notice Last snapshotted timestamp
+    uint256 public lastSnapshotTimestamp;
+
+    /// @notice Amount of revenue token in the ancestors vault
+    mapping(address token => uint256 amount) public ancestorsVaultAmount;
+
+    /// @notice Indicates if a given ancestor address has already claimed
+    mapping(address ancestorIpId => bool) public isClaimedByAncestor;
+
+    /// @notice Amount of revenue token in the claim vault
+    mapping(address token => uint256 amount) public claimVaultAmount;
+
+    /// @notice Amount of tokens claimable at a given snapshot
+    mapping(uint256 snapshotId => mapping(address token => uint256 amount)) public claimableAtSnapshot;
+
+    /// @notice Amount of unclaimed tokens at the snapshot
+    mapping(uint256 snapshotId => uint32 tokenAmount) public unclaimedAtSnapshot;
+
+    /// @notice Indicates whether the claimer has claimed the revenue tokens at a given snapshot
+    mapping(uint256 snapshotId => mapping(address claimer => mapping(address token => bool)))
+        public isClaimedAtSnapshot;
+
+    /// @notice Royalty tokens of the vault
+    EnumerableSet.AddressSet private _tokens;
+
+    /// @notice Constructor
+    /// @param royaltyPolicyLAP The address of the royalty policy LAP
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address royaltyPolicyLAP) {
+        if (royaltyPolicyLAP == address(0)) revert Errors.IpRoyaltyVault__ZeroRoyaltyPolicyLAP();
+        ROYALTY_POLICY_LAP = IRoyaltyPolicyLAP(royaltyPolicyLAP);
+        _disableInitializers();
+    }
+
+    // TODO: adjust/review for upgradeability
+    /// @notice Initializer for this implementation contract
+    /// @param name The name of the royalty token
+    /// @param symbol The symbol of the royalty token
+    /// @param supply The total supply of the royalty token
+    /// @param unclaimedTokens The amount of unclaimed royalty tokens reserved for ancestors
+    /// @param ipIdAddress The ip id the royalty vault belongs to
+    function initialize(
+        string memory name,
+        string memory symbol,
+        uint32 supply,
+        uint32 unclaimedTokens,
+        address ipIdAddress
+    ) external initializer {
+        if (ipIdAddress == address(0)) revert Errors.IpRoyaltyVault__ZeroIpId();
+
+        ipId = ipIdAddress;
+        lastSnapshotTimestamp = block.timestamp;
+        unclaimedRoyaltyTokens = unclaimedTokens;
+
+        _mint(address(this), unclaimedTokens);
```

Comment:
who can call this initialize function?  is it any risk of excessive mint here?

Reply:
do you need to check if unclaimedTokens < TOTAL_SUPPLY?

Reply:
`initalize` is called in [RoyaltyPolicyLAP.sol](https://github.com/storyprotocol/protocol-core-v1/blob/f89fbb9d1cfdbdebb6f113c641646bd7112ad167/contracts/modules/royalty/policies/RoyaltyPolicyLAP.sol#L241-L243)

`royaltyStack` in RoyaltyPolicyLAP is passed as `unclaimedTokens`, which is meant to be claimed by parents. The rest `supply - royaltyStack` is sent to the IP Account directly on creation.

Since `_initPolicy` in RoyaltyPolicyLAP validates that `royaltyStack` is at max `TOTAL_RT_SUPPLY` (via `_getNewAncestorsData` before initialize IpRoyaltyVault), we don't need to check the inequality.

We may want to gate the `initialize` call to only the RoyaltyPolicyLAP.

Reply:
Initialize can only be called once otherwise it reverts with `Initializable: contract is already initialized`. Given it's called by RoyaltyPolicyLAP.sol and then no one else can call it after that we may not need to gate the function call.

Reply:
Can Initialize be called 2nd time after upgrade?  If it is only called during deployment, the risk seems low. But if we only allow it be called by Royal Module, do you need to limit it?

Reply:
resolve it for now, we can do offline communication for the clarification.